source/speech.py

```diff
@@ -137,52 +131,76 @@ def getCurrentLanguage():
-	global beenCanceled, _speakSpellingGenerator
-	import speechViewer
-	if speechViewer.isActive:
-		speechViewer.appendText(text)
-	if speechMode==speechMode_off:
-		return
-	elif speechMode==speechMode_beeps:
-		tones.beep(config.conf["speech"]["beepSpeechModePitch"],speechMode_beeps_ms)
-		return
-	if isPaused:
-		cancelSpeech()
-	beenCanceled=False
+def getSpeechForSpelling(text, locale=None, useCharacterDescriptions=False):
 	defaultLanguage=getCurrentLanguage()
 	if not locale or (not config.conf['speech']['autoDialectSwitching'] and locale.split('_')[0]==defaultLanguage.split('_')[0]):
 		locale=defaultLanguage
 
 	if not text:
 		# Translators: This is spoken when NVDA moves to an empty line.
-		return getSynth().speak((_("blank"),))
+		yield _("blank")
+		return
 	if not text.isspace():
 		text=text.rstrip()
-	if _speakSpellingGenerator and _speakSpellingGenerator.gi_frame:
-		_speakSpellingGenerator.send((text,locale,useCharacterDescriptions))
-	else:
-		_speakSpellingGenerator=_speakSpellingGen(text,locale,useCharacterDescriptions)
-		try:
-			# Speak the first character before this function returns.
-			next(_speakSpellingGenerator)
-		except StopIteration:
-			return
-		queueHandler.registerGeneratorObject(_speakSpellingGenerator)
+
+	synth = getSynth()
+	synthConfig=config.conf["speech"][synth.name]
+	charMode = False
+	textLength=len(text)
+	count = 0
+	localeHasConjuncts = True if locale.split('_',1)[0] in LANGS_WITH_CONJUNCT_CHARS else False
+	charDescList = getCharDescListFromText(text,locale) if localeHasConjuncts else text
+	for item in charDescList:
+		if localeHasConjuncts:
+			# item is a tuple containing character and its description
+			char = item[0]
+			charDesc = item[1]
+		else:
+			# item is just a character.
+			char = item
+			if useCharacterDescriptions:
+				charDesc=characterProcessing.getCharacterDescription(locale,char.lower())
+		uppercase=char.isupper()
+		if useCharacterDescriptions and charDesc:
+			char=charDesc[0] if textLength>1 else u"\u3001".join(charDesc)
+		else:
+			char=characterProcessing.processSpeechSymbol(locale,char)
+		if uppercase and synthConfig["sayCapForCapitals"]:
+			# Translators: cap will be spoken before the given letter when it is capitalized.
+			char=_("cap %s")%char
+		if uppercase and synth.isSupported("pitch") and synthConfig["capPitchChange"]:
+			yield PitchCommand(offset=synthConfig["capPitchChange"])
+		if config.conf['speech']['autoLanguageSwitching']:
+			yield LangChangeCommand(locale)
+		if len(char) == 1 and synthConfig["useSpellingFunctionality"]:
+			if not charMode:
+				yield CharacterModeCommand(True)
+				charMode = True
+		elif charMode:
+			yield CharacterModeCommand(False)
+			charMode = False
+		if uppercase and  synthConfig["beepForCapitals"]:
+			yield BeepCommand(2000, 50)
+		yield char
+		if uppercase and synth.isSupported("pitch") and synthConfig["capPitchChange"]:
+			yield PitchCommand()

```

Comment:
Since this PitchCommand needs to be tied to the first one, it would be nice to use some RAII like idiom to ensure that is the case. At the moment this seems to work a bit like a state machine, and is vulnerable to later modifications that may change the control flow but don't keep in mind the same cases as when originally developed.